Add Go 1.20 errors.Join support

dhartunian opened the pull request:
This PR contains the `Join` implementation from #106 and adds
some adjustments to fit the rest of the library's style.

We move the newly added `Join` implementation that mirrors
the go 1.20 stdlib function into a separate package like the rest of
our custom error types.

Some simple unit tests are added and `Join` wrappers are also
integrated into the datadriven formatting test. Our existing format
code for multi-cause errors is compatible with this new type which
allows us to remove the custom formatter from the implementation.

----
Please review last 2 commits. First one is #115.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/errors/118)
<!-- Reviewable:end -->

knz reviewed (commented):
May I suggest:

- move the new join.go to its sub-package `join` immediately in the 1st commit.
- ensure the `joinError` type also implements `SafeFormatError`. Then express Error/Format using that, using the model already set by other types.
- in the 2nd commit, make the top-level join `Join` point to a wrapper in package `errutil` that follows the example of `Wrap` and also automatically adds a call to `WithStack`.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/errors/118#-:-Nc2YUt0C2jriEwfuMjX:bqe630g)* status: 0 of 18 files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

knz reviewed (commented):
(maybe my last consideration is not so great - the opaque type's support for redactable strings is janky)

If you do not do that, consider adding a decoder for the new joinError type.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/errors/118#-:-Nc2_0o75KhePfrhe3M4:bjam73g)* status: 0 of 18 files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

dhartunian commented:
@knz PTAL

- added custom encoder/decoder for Join (based on API extended in latest update to #115)
- merged the two commits into 1 to put the Join impl where it needs to be
- added custom formatter
- added public API wrapper with stack

knz commented:
CI is trying to tell you something.

I'd say let's merge #115 then you can rebase this which will make the review easier.

dhartunian commented:
> CI is trying to tell you something.

lol fixed 😄

knz reviewed (commented):
> Reviewed 17 of 18 files at r5, 8 of 8 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/errors/118)* status: all files reviewed, 2 unresolved discussions (waiting on @dhartunian)

___
*[`go.mod` line 3 at r6](https://reviewable.io/reviews/cockroachdb/errors/118#-NcSSjR74SHb2LCS8bnY:-NcSSjR74SHb2LCS8bnZ:bh6jrn2) ([raw file](https://github.com/cockroachdb/errors/blob/5a7989c9c0a05823805bd6e01580c2fdba48405d/go.mod#L3)):*
> ```XML
> module github.com/cockroachdb/errors
> 
> go 1.18
> ```

I'm not sure you wanted to change this?
___
*[`join/join.go` line 76 at r6](https://reviewable.io/reviews/cockroachdb/errors/118#-NcSSwDK4QyJRMtMwOSY:-NcSSwDK4QyJRMtMwOSZ:bg82ue7) ([raw file](https://github.com/cockroachdb/errors/blob/5a7989c9c0a05823805bd6e01580c2fdba48405d/join/join.go#L76)):*
> ```Go
> 
> func (e *joinError) SafeFormatError(p errbase.Printer) error {
> 	p.Print(e.Error())
> ```

Please do it the other way around: put the logic in here, then do `Error() string { return redact.StringWithoutMarkers(e) }`

Otherwise you'd be throwing redactability out of the window.

(you can see the effects of this in the format tests)


<!-- Sent from Reviewable.io -->

dhartunian reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/errors/118)* status: 13 of 25 files reviewed, 2 unresolved discussions (waiting on @knz)

___
*[`go.mod` line 3 at r6](https://reviewable.io/reviews/cockroachdb/errors/118#-NcSSjR74SHb2LCS8bnY:-NcSVx3i-2aV8m6ZI4Hr:bp69eqp) ([raw file](https://github.com/cockroachdb/errors/blob/5a7989c9c0a05823805bd6e01580c2fdba48405d/go.mod#L3)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

I'm not sure you wanted to change this?
</blockquote></details>

whoops. done.
___
*[`join/join.go` line 76 at r6](https://reviewable.io/reviews/cockroachdb/errors/118#-NcSSwDK4QyJRMtMwOSY:-NcXNvM47Qjvsg105y8E:b-7rl0jd) ([raw file](https://github.com/cockroachdb/errors/blob/5a7989c9c0a05823805bd6e01580c2fdba48405d/join/join.go#L76)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

Please do it the other way around: put the logic in here, then do `Error() string { return redact.StringWithoutMarkers(e) }`

Otherwise you'd be throwing redactability out of the window.

(you can see the effects of this in the format tests)
</blockquote></details>

done.


<!-- Sent from Reviewable.io -->

knz reviewed (approved):
:100: 

> Reviewed 12 of 12 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/errors/118#-:-NcXp3DTCWPl_QXyPYTy:bmi8ppe)* status: all files reviewed, 1 unresolved discussion (waiting on @dhartunian)



<!-- Sent from Reviewable.io -->
